### ISSUE TITLE
Mypy 1.6 Release Planning

### ISSUE THREAD

**Issue by koogoro** [python/mypy#15919]
I'm planning to make the 1.6 release by the end of August.

Please post any issues you'd like to see fixed in the release here, or any PRs you'd like to have merged. I'll make the release branch `release-1.6`.

**Comment by koogoro** [comment]
I've run into a regression: https://github.com/python/mypy/issues/15932.

**Comment by tyralla** [comment]
If this is the right place to bring attention to old, so far undiscussed pull requests: #14440

**Comment by AlexWaygood** [comment]
It would be good to get #15960 (or an equivalent fix) in before the release. It fixes some recent stubtest regressions that I introduced.

**Comment by ilevkivskyi** [comment]
@koogoro I have a suggestion on the blog post: Could you please mention there that `--new-type-infernece` will be on by default in 1.7?

**Comment by koogoro** [comment]
The branch has been created. The new functionality in the `attrs` plugin to cover `fields` might cause some errors to pop up for users of `attrs` but this isn't really a regression and I don't think the primer noticed anything, so I've marked this as a feature request: https://github.com/python/mypy/issues/15980.

**Comment by koogoro** [comment]
@AlexWaygood That commit didn't make it onto the branch, and it also doesn't cherry-pick cleanly. Does that mean that the change that introduced the issue didn't make it onto the branch, or would it be good to include multiple more commits on the release branch?

**Comment by cdce8p** [comment]
@koogoro Thanks for creating the release branch. I noticed that it branched of over a week ago. Was that intentional? It's missing a lot of the recent PRs especially from @ilevkivskyi. If you're not too far along with testing it, might I suggest hard reseting the branch to the current master commit? 010da0b

https://github.com/python/mypy/commits/master
https://github.com/python/mypy/commits/release-1.6

**Comment by koogoro** [comment]
@cdce8p The wheels build failed for the latest commit, so I'll use the one before it, d7b24514d7301f86031b7d1e2215cf8c2476bec0. Let me quickly check if there are any large problems before hard resetting it.

**Comment by AlexWaygood** [comment]
> @AlexWaygood That commit didn't make it onto the branch, and it also doesn't cherry-pick cleanly. Does that mean that the change that introduced the issue didn't make it onto the branch, or would it be good to include multiple more commits on the release branch?

I was hoping https://github.com/python/mypy/commit/0b303b53479897e24d57affef6a8cdfffbd08e3d would make it into the release, and if it does, it's important https://github.com/python/mypy/commit/d7b24514d7301f86031b7d1e2215cf8c2476bec0 does also, as it fixes some regressions introduced by https://github.com/python/mypy/commit/0b303b53479897e24d57affef6a8cdfffbd08e3d. But if neither commit makes it in, that's also okay, it can wait :)

It would be pretty useful if https://github.com/python/mypy/commit/48835a362d86eb9964b0350e4453daf14c76fe8e could be included in the release, though — this one fixes a crash that's blocking some typeshed updates.

**Comment by koogoro** [comment]
I did find one regression in d7b24514d7301f86031b7d1e2215cf8c2476bec0: https://github.com/python/mypy/issues/15984. This regression's new in between that and the current release point, maybe due to some `ParamSpec` change between the two?

**Comment by hauntsaninja** [comment]
It would be good to cherrypick https://github.com/python/mypy/pull/16004. You may also need https://github.com/python/mypy/pull/15992 for CI to work

**Comment by hauntsaninja** [comment]
We should cherrypick https://github.com/python/mypy/pull/16013 in case we don't have a release before Python 3.12 is out

**Comment by koogoro** [comment]
Cherry-picked 0ae0c75 as a fix for #15984 and 379b52f as a fix for #15990. I'll cherry-pick #16004 and #16013 once they've been merged.

**Comment by Hnasar** [comment]
Can we please get https://github.com/python/mypy/commit/010da0b2f48dc92be2f79495fd4551c92351868f for https://github.com/python/mypy/issues/15975  ?

**Comment by hauntsaninja** [comment]
Both #16004 and #16013 are merged.

Quick note about #16004: it's extremely commonly requested, so we do need to do it, but it might be a little annoying to upgrade for projects that use type ignore with error codes and that use `--warn-unused-ignore`. If we're concerned about that, some options are https://github.com/python/mypy/pull/16004#issuecomment-1700339306. If we want to figure it out in release 1.7, then we should maybe revert #16004, #15840, #14740 on the 1.6 branch.

**Comment by cdce8p** [comment]
Small change to better handle the `--new-type-inference` flag. It would make sense to cherry-pick that as well IMO.
- https://github.com/python/mypy/pull/16059

**Comment by Dreamsorcerer** [comment]
> it might be a little annoying to upgrade for projects that use type ignore with error codes and that use `--warn-unused-ignore`

As a maintainer of 1 of the projects in the mypy primer diff, I'm fine with making these updates manually. More precise ignores seems like a good thing. We often have to make updates to remove ignores etc. as mypy gets better anyway.

**Comment by mr-c** [comment]
Getting https://github.com/python/mypy/pull/16124 in for mypy 1.6 would fix a regression that prevents `cwltool` (59k downloads a month on PyPI, total 4M PyPI, 475k conda) from upgrading mypy since 1.4.0.

**Comment by huix11** [comment]
11123

**Comment by hauntsaninja** [comment]
@koogoro let me know if there's something I can do to help the release!

**Comment by ilevkivskyi** [comment]
@JukkaL @koogoro What is the situation with the release? What is currently the blocker?

**Comment by JukkaL** [comment]
I will work on finishing the release. I hope to get it out by early next week. I'm planning to start working on the 1.7 release soon afterwards.

**Comment by JukkaL** [comment]
I worked on the release blog post today, and hope to complete it tomorrow. After the blog post is done, we should be good to publish the release. If anybody has any other things needed before the release, please leave a comment.

**Comment by ilevkivskyi** [comment]
Just one small reminder on the blog post:
> I have a suggestion on the blog post: Could you please mention there that `--new-type-infernece` will be on by default in 1.7?

**Comment by cdce8p** [comment]
_I mentioned it earlier but AFAICT it hasn't been cherry-picked yet. https://github.com/python/mypy/issues/15919#issuecomment-1709104729_

Especially for the `--new-type-inference` flag, cherry-picking this PR would make sense IMO.

- https://github.com/python/mypy/pull/16059

**Comment by mr-c** [comment]
@JukkaL please also cherry pick https://github.com/python/mypy/pull/16124/ as per https://github.com/python/mypy/issues/15919#issuecomment-1722214168

**Comment by ilevkivskyi** [comment]
I think these two mentioned by @hauntsaninja are important
> Both https://github.com/python/mypy/pull/16004 and https://github.com/python/mypy/pull/16013 are merged and should be cherry picked.

In particular, we must fail gracefully on Python 3.12 that was just released.

**Comment by JukkaL** [comment]
Cherry-picked these PRs to the release branch:
* #16059 
* #16124 
* #16004 
* #16013

**Comment by JukkaL** [comment]
And the release is out: https://mypy-lang.blogspot.com/2023/10/mypy-16-released.html

**Comment by JukkaL** [comment]
I will update the website tomorrow, since it's getting late here.

**Comment by AlexWaygood** [comment]
A new crash with mypy 1.6, that I hit when trying to bump typeshed's mypy pin:

- https://github.com/python/mypy/issues/16245

**Comment by intgr** [comment]
@ymyzk Can you please add mypy 1.6.0 to mypy-play.net?

**Comment by ymyzk** [comment]
@intgr Yes, thanks for the reminder. I added mypy 1.6.0 and updated mypy-play.net https://github.com/ymyzk/mypy-playground/pull/488

**Comment by intgr** [comment]
I experienced two regressions with mypy 1.6.0, both involving `ParamSpec`:

* #16246
* #16247

**Comment by ilevkivskyi** [comment]
An obvious candidate for 1.6.1 https://github.com/python/mypy/pull/16251 (fixes two closely related crash regressions).

**Comment by AlexWaygood** [comment]
Another new crash in mypy 1.6 (that still reproduces with mypy `master` and the latest version of the `release-1.6` branch):

- #16257

**Comment by AlexWaygood** [comment]
#16258 fixes a regression of sorts, so _might_ be worth including in a patch release? But the regression was a new in mypy 1.5 rather than mypy 1.6.

**Comment by ilevkivskyi** [comment]
Also CP for 1.6.1 https://github.com/python/mypy/pull/16259

**Comment by intgr** [comment]
FYI mypy website https://mypy-lang.org/ does not display the 1.6 release in "What's new"

**Comment by hauntsaninja** [comment]
I cherrypicked these to release-1.6 branch and will cut the release later today:
```
6b6504eb8 Fix crash on ParamSpec unification (for real) (#16259)
eb81e63e9 Fix crash on ParamSpec unification (#16251)
```

**Comment by hauntsaninja** [comment]
https://pypi.org/project/mypy/1.6.1/ is now a thing

I also left some notes at https://github.com/python/mypy/wiki/Release-Process#making-a-patch-release so I don't have to grep through my shell history next time to see if I forgot something
